### DISCUSSION TITLE
Multiple VI Outputs.

### DISCUSSION THREAD

**Discussion by lesbrunn** [apohl79/audiogridder#584] [Q&A]
Hello guys,

For about a week, I've been trying to understand AG to the point where I can use it professionally. There are still a few issues. Many VI's such as Kontakt have multiple outputs which can be separated in most DAWs. It seems AG does not recognize multi outputs? Since it has 16 MIDI channels, one should be able to have a different MIDI channel feeding different outputs so one can have separation in the DAW. Am I doing something wrong? Any light on this would be appreciated, since this is a major thing in my workflow. 

Thanks,
Leslie.

**Comment by lesbrunn** [comment]
I guess the only way around this presently is to have many instances of AG with the VI loaded, each instance assigned to a different output and to a different DAW track. That sort of defeats the elegance of having just one VI loaded that addresses its multi outputs while allowing assignment to different DAW tracks. Is there any hope of having this issue looked into?

**Reply by apohl79** [reply]
I'm sure AG is lacking support for things in the instruments area. As I'm not using this myself, I'm likely lacking an understanding of what is needed by composers. So please describe these things the way you would expect them to work, and we can create feature request tickets for future releases.

**Comment by lesbrunn** [comment]
Hi Andreas,
Thank you for your time and enthusiasm. You are doing a tremendous job, and even though AG is not fully mature, I am beginning to like it very much after the initial teething problems of trying to understand the software and the way it works. I really battled long and hard before I could get it working smoothly. That's not saying that AG is difficult to use, but the instructions are rather cryptic, and one has to experiment before understanding fully dawns.
Let me see if I can describe this output issue in a way that would make sense. Although all DAW's work essentially the same way, I am more familiar with Pro Tools so I will approach my description from that viewpoint. 
Let me give an example with Native Instruments Kontakt, which is a very popular sample player. Kontakt is a multitimbral instrument, meaning that you can load several instruments into it at the same time to be played from the DAW. It is also a multi-output instrument, allowing you to assign different instruments to different outputs. Thus you can have, say, a piano, a bass, a trumpet and a drum kit loaded into one instance of Kontakt. Each instrument can be discrete, with its own midi channel and output assignment. Moreover, even the drum kit (usually regarded as one instrument) can be organized to output its different components (snare, kick, toms etc,) to their own discrete outputs.
The DAW is ready to receive all this info, and you can set up different tracks for each instrument on different midi channels. You can have the piano on midi ch1, the bass on ch2 and so on. More or less, you set up the DAW to agree with the assignments you have already made in Kontakt. 
However, in AG, each instrument only outputs audio through its main stereo outputs. Of course, this is OK for most synths that only have a stereo output. Some multitimbral VI's can contain many instruments, but have all of them going out of one stereo output. Some may have two or three different stereo outputs and some can have 32 or even 64 different stereo outputs. These are usually flexible in their configuration, so you can have the outputs configured as separate L and R (mono outputs) or combinations of mono and stereo outputs. Most DAW's can accommodate both mono and stereo inputs and outputs. Usually, at the input configuration point in the DAW, you are given a list of the outputs that the plugin (Kontakt in our example) presents to the DAW. You can then assign them to inputs in the DAW and have your different instruments channel audio into discrete DAW tracks.
However, with AG, when a multi-output plugin is instantiated in Pro Tools, the input configurator is greyed out. You have access to only the main stereo out of the plugin which automatically feeds the track on which AG is presenting the plugin.
 A common scenario would be for AG to be instantiated on one track, and have several adjacent tracks assignable to the various outputs that Kontakt presents through AG, without actually having AG instantiated on those tracks. In other words, a single instance of AG on one track would allow the DAW access to all the output parameters that the VI offers, and would allow their configuration across several other tracks.
This has been a clumsy description at best, and I hope it hasn't been too confusing. Please feel free to ask me specific questions in areas where you need more clarity. Thank you, Andreas.
Best regards,Leslie.

    On Tuesday, May 25, 2021, 07:11:00 PM UTC, Andreas Pohl ***@***.***> wrote:  
 
 


I'm sure AG is lacking support for things in the instruments area. As I'm not using this myself, I'm likely lacking an understanding of what is needed by composers. So please describe these things the way you would expect them to work, and we can create feature request tickets for future releases.

**Reply by apohl79** [reply]
Started to look at this now. Would there be any issue with AG reporting separate stereo channels to the DAW or would it be necessary to have mono channels?

This is how it would look like in cubase (8 stereo pairs):

![image](https://user-images.githubusercontent.com/1129160/123293251-20209300-d514-11eb-9ea3-4cbfdcd1f84c.png)

**Reply by apohl79** [reply]
also, looking at kontakt, more channels are needed as well?

**Reply by apohl79** [reply]
Maybe the instrument output channel bus config should be:

1: Main L/R (2 channels)
2: Mono 1 (1 channel)
...
15: Mono 14 (1 channel)

Could also 64 channels in total, if that makes sense.

**Reply by emack999** [reply]
Would definitely plan for a 64 channel configuration, as Kontakt is pretty common.

**Reply by lesbrunn** [reply]
> Maybe the instrument output channel bus config should be:
> 
> 1: Main L/R (2 channels)
> 2: Mono 1 (1 channel)
> ...
> 15: Mono 14 (1 channel)
> 
> Could also 64 channels in total, if that makes sense.

Hi Andreas,

Great to see you are working to improve AG. Each stereo output is usually also configurable as 2 mono outputs. L & R. One has the option to choose stereo or mono.

The only thing that prevents me from using AG is the constant dropouts in the audio, almost as if AG is struggling hard to keep up with the audio, and fails sometimes. When this happens, I get huge cpu spikes in Pro Tools (AAE 9173). The spikes go away when I remove AG. This happens so often that I can't use it on serious projects. AG is such a great concept; there is nothing quite like it. Except VE Pro, which I also use, and which works smoothly for me on large projects. The advantage AG has is that you see the plugins in your DAW as if it's your DAW doing everything. Sometimes with only 2 instances of AG running VI's, I get 100% cpu usage in Pro Tools, where I previously had only 10 or 15% usage. This causes Pro Tools to crash.

**Reply by apohl79** [reply]
Need to investigate that further. Once I get to it, I might have more questions to be able to reproduce this in my environment.

**Reply by lesbrunn** [reply]
> Need to investigate that further. Once I get to it, I might have more questions to be able to reproduce this in my environment.

Great. More grease!

**Reply by dhstsw** [reply]
Hi,
Just noticed this.
I guess the best option would be the initial one, having stereo couples.
Or, in alternative, being able to configure it (stereo or mono).
But most of plugins reason in stereo busses.

C.

**Reply by apohl79** [reply]
implemented it format dependent now. For protools/aax stereo buses seemed best. VST has a stereo main plus mono buses, and AU needed a JUCE patch and something else...

**Reply by lesbrunn** [reply]
Well done, Andreas. Would like to check it out, but it seems you haven't posted it yet? Looking forward to when I can use it solidly with multiple instances in a large Pro Tools session without CPU spikes and crashes...

**Reply by Dumdididum** [reply]
I have loaded the NI's Komplete Kontrol into Cakewalk and it loads 16 individual tracks. Opening up the KK snyth on the first track, you can load a copy of Audio Gridder (e47) which you only need to do on the first track. I've continued by loading drum synths that have multiple outputs into AG on the slave PC, and so long I remember to assign the kick, snare, hihat, toms, etc. to separate tracks, they all show up on in Cakewalk on the different tracks that KK has loaded! 

Pretty cool!

Except, when I do a check on CPU/MEM usage (Task Manager), it seems KK is using a higher percent of power/mem then it does if I just load the drum VSTs directly in to Cakewalk, without using KK with AG.
I'm assuming that KK uses a lot of CPU/MEM when it is loaded.

If anybody knows of a similar plugin like KK that is capable of loading VSTi and has multiple outputs (tracks) but has less of a CPU/MEM load, it would be nice to share it. It may work just as KK does in giving separate tracks for each output on your main DAW.

**Reply by lesbrunn** [reply]
After not using AG for some time, I tested the new beta yesterday. I wanted to check out the multiple outputs, so I loaded SSD5, which I use quite often. I successfully used up to 8 stereo outputs with 2 blocks at a buffer of 64 in Pro Tools. Latency wasn't too bad, but I had several stalls and crashes, and I had to restart AG several times. The main issue, though, is that with only SSD5 in PT at the above-mentioned settings, PT's CPU meter was very unstable and sometimes shot up to 48%. Disabling AG brought the meter down to 1%. Loading SSD5 with the same 8 outputs in PT without AG gave me a reading of only 4% at the highest peak. As solid as the AG concept is, the purpose of reducing the DAW's CPU load isn't yet happening. There's still a lot of work to be done for AG to lift the load off PT. To be fair, though, I haven't yet tested AG with other VI's. I intend to do so when I have some time, and I'll post my findings.

**Reply by lesbrunn** [reply]
> I have loaded the NI's Komplete Kontrol into Cakewalk and it loads 16 individual tracks. Opening up the KK snyth on the first track, you can load a copy of Audio Gridder (e47) which you only need to do on the first track. I've continued by loading drum synths that have multiple outputs into AG on the slave PC, and so long I remember to assign the kick, snare, hihat, toms, etc. to separate tracks, they all show up on in Cakewalk on the different tracks that KK has loaded!
> 
> Pretty cool!
> 
> Except, when I do a check on CPU/MEM usage (Task Manager), it seems KK is using a higher percent of power/mem then it does if I just load the drum VSTs directly in to Cakewalk, without using KK with AG.
> I'm assuming that KK uses a lot of CPU/MEM when it is loaded.
> 
> If anybody knows of a similar plugin like KK that is capable of loading VSTi and has multiple outputs (tracks) but has less of a CPU/MEM load, it would be nice to share it. It may work just as KK does in giving separate tracks for each output on your main DAW.

Reason is a very solid choice. Light on CPU, with tons of instruments and effects. Also loads VST3 plugins.

**Reply by lesbrunn** [reply]
Sorry, Reason only loads VST3 plugins in standalone mode. But check it out, The Reason Rack plugin has a lot to offer and is both CPU efficient and very usable. Blue Cat Patchwork is a stable plugin host, and so is Vienna Ensemble Pro. I use both of them to host VSTi's in Pro Tools.
